### PR TITLE
Consolidate Results validation + add Results to Run status

### LINKED ISSUE #138: Concise Yaml for pipeline declaration.
## Expected Behavior
user should be able to write minimal lines of yaml to declare a pipeline.

## Actual Behavior
we end up writing a number of redundant fields like "name" in inputSourceBindings might not be required.

## Steps to Reproduce the Problem

  1. Try writing a sample pipeline
  2. Remove repeated fields.
  3.

### LINKED ISSUE #146: TaskRun should have `Results` in `Status`
## Expected Behavior
When `TaskRun` executes, it will upload its logs and results to the locations in `PipelineParams`. A user should be able to look at the Status of a `TaskRun` and see exactly where those logs ended up (i.e. the whole path).

This path should be determined by the controller, e.g. a folder in the path named after the TaskRun and/or timestamp, and provided in the `Status` of the `TaskRun`.

## Actual Behavior
The `TaskRun` _Spec_ (not `Status`!) has [Results](https://github.com/knative/build-pipeline/blob/master/pkg/apis/pipeline/v1alpha1/taskrun_types.go#L32) which in a previous iteration of our examples were expected to be provided manually:

https://github.com/knative/build-pipeline/blob/dcc2cb189f268c1a9651499c1c1f6ce5f76b1e02/examples/invocations/run-kritis-test.yaml#L24-L36

I deleted this from the example b/c I was thinking that referencing the PipelineParams was enough.

## Additional Info
Note #107 is about actually making this work, this is more about the API.

### PR DESCRIPTION
**Pull request by bobcatfish**
When a user kicks off a run, they will provide an endpoint to upload logs to (initial implementation will be in #107). The corresponding fields in `status` will indicate where the logs actually got uplaoded to.

Once we actually get to #107, and especially once we start supporting endpoints other than GCS, we may find this isn't useful and remove it.

As a bonus:
* Remove `name` field from `Results`
  In the interests in keeping our spec as lean and simple as it can possibly be, at the moment there is no reason why our Results config needs to have a name, so let's remove it. This is the last unnecessary field we have in our yamls that I'm aware of, so this fixes #138
* Consolidate logic for handling results config for PipelineRun + TaskRun
  It turned out TaskRun and PipelienRun were using slightly different logic for validating the results config, even tho it should be the same for both. Now they share the logic and in both cases the config is optional.
* Simplify results config by removing `log` key
  Since currently the only kind of result is `log`, we can remove that key. We change change this if we decide to have other kinds of results later.

Fixes #146
Fixes #138

### CODE REVIEW

File: pkg/apis/pipeline/v1alpha1/pipelinerun_validation.go [hunk 1]
```diff
@@ -46,33 +44,19 @@ func (ps *PipelineRunSpec) Validate() *apis.FieldError {
 
 	if ps.Results != nil {
 		// Results.Logs should have a valid URL and ResultTargetType
-		if err := validateURL(ps.Results.Logs.URL, "pipelinerun.spec.Results.Logs.URL"); err != nil {
+		if err := validateURL(ps.Results.URL, "pipelinerun.spec.Results.URL"); err != nil {
 			return err
 		}
-		if err := validateResultTargetType(ps.Results.Logs.Type, "pipelinerun.spec.Results.Logs.Type"); err != nil {
+		if err := validateResultTargetType(ps.Results.Type, "pipelinerun.spec.Results.Type"); err != nil {
 			return err
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/pipelinerun_validation.go]: whoops, forgot to delete this section - nice catch coverage tool XD

**Comment by abayer** [comment]
> Since currently the only kind of result is log, we can remove that key. We change change this if we decide to have other kinds of results later.

Yeah, that seems fine to me for now, but I know I'm going to want to eventually have more possible types.

File: pkg/apis/pipeline/v1alpha1/pipelinerun_types.go [hunk 2]
```diff
@@ -137,6 +111,8 @@ type PipelineTriggerRef struct {
 // PipelineRunStatus defines the observed state of PipelineRun
 type PipelineRunStatus struct {
 	Conditions duckv1alpha1.Conditions `json:"conditions"`
+	// In #107 should be updated to hold the location logs have been uploaded to
+	Results Results `json:"results"`
```
Comment (pivotal-nader-ziada) [inline pkg/apis/pipeline/v1alpha1/pipelinerun_types.go]: The results field in the pipelineRun is optional but here in the status it not, what happens if there is a target url for the results?

File: pkg/apis/pipeline/v1alpha1/taskrun_types.go [hunk 3]
```diff
@@ -103,6 +104,9 @@ type TaskRunStatus struct {
 	// +optional
 	Conditions duckv1alpha1.Conditions `json:"conditions,omitempty"`
 
+	// In #107 should be updated to hold the location logs have been uploaded to
+	Results Results `json:"results"`
```
Comment (pivotal-nader-ziada) [inline pkg/apis/pipeline/v1alpha1/taskrun_types.go]: same comment as above, the results field is optional in the taskRun, but the field in the status is not

File: pkg/apis/pipeline/v1alpha1/pipelinerun_types.go (hunk 2 above)
Comment (bobcatfish) [reply pkg/apis/pipeline/v1alpha1/pipelinerun_types.go]: we can make this optional too!

**Comment by bobcatfish** [comment]
>  I know I'm going to want to eventually have more possible types.

Super cool @abayer ! If you feel like writing up an issue or doc with some details about what types you'd like to see that would be super cool! No worries if it's too premature. and p.s. @tejal29 would be a good contact to continue to discuss the results design

**Comment by bobcatfish** [comment]
Ready for another look @pivotal-nader-ziada !

**Comment by bobcatfish** [comment]
> @pivotal-nader-ziada: Bad category. Please see https://api.thecatapi.com/api/categories/list

nooooo

/meow space

**Pull request merged**